lib/pbench/server/templates.py

```diff
@@ -72,6 +103,66 @@ def add_mapping(self, name: str, body: Dict[AnyStr, Any]):
         self.json["properties"][name] = body
 
 
+class JsonToolFile(JsonFile):
+    """
+    Extend the JSON template file handling to provide merging JSON
+    tool templates with a base skeleton.
+    """
+
+    def load(self):
+        """
+        Tool mapping files define a sub-document in the Elasticsearch template
+        specific to the tool (e.g., "iostat") which exists within a common
+        base skeleton. The merge of the two includes elevating the "_meta"
+        version from the sub-document into the base document.
+
+        Here we extend the JsonFile load method to remove the sub-document
+        metadata and save it to be spliced into the base document by merge.
+        """
+        super().load()
+        self.meta = self.json["_meta"]
+        del self.json["_meta"]
+
+    def get_version(self) -> str:
+        """
+        Parse the metadata version information from the mapping file. This is
+        done on standalone template files and on "tool" template files, but
+        not on the "base" skeleton files shared by the tool templates. The
+        context is under control of the TemplateFile class load methods.
+
+        This produces a similar result to the superclass method; except that
+        JsonToolFile has already sequestered the version metadata to prepare
+        for merging with the skeleton.
+
+        Raises:
+            MappingFileError: Mapping file doesn't contain proper version
+
+        Returns:
+            Template file metadata version
+        """
+        try:
+            version = self.meta["version"]
+        except KeyError as e:
+            raise MappingFileError(f"mapping missing {e} in {self.file}: {self.meta}")
+        return version
```

Comment:
I kind of wish that this were refactored into a helper function which was passed the JSON blob as a parameter -- that is, it'd be up to the caller to decide whether it is in `self.meta` or `self.json["_meta"]`, and it'd up to the subroutine to extract it and deal with the error.

It's not quite as clean/effective as I'd like it to be, because the access to the "_meta" key is done outside the helper, but it would be DRYer.

Reply:
Yeah:  hmm....  I figured I would offer.